ksql-engine/src/main/java/io/confluent/ksql/function/udtf/Cube.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.function.udtf;
+
+import io.confluent.ksql.util.KsqlConstants;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+@UdtfDescription(name = "cube", author = KsqlConstants.CONFLUENT_AUTHOR,
```

Comment:
could we call this `cube_explode` , so that we could reserve `cube` as a keyword if we choose to add real support via grouping sets down the line?

Reply:
This will also make it clearer that this is a UDTF, since explode is a well understood UDTF already across many projects.

Reply:
Agree on reserving `cube` name for the real thing. Perhaps this fn is more of a `permute` ? Either way, it feels like kind of a hack for simply enhancing the recently-added `explode` fn to have a variant which takes multiple input column names (variadic version) or another override which takes a list of columns ? Adding a whole new fn to achieve that feels like an unnecessary cognitive burden ?

Reply:
@blueedgenick The cube udtf actually does more than the explode one. Explode creates as many rows as tuples in the array. Cube creates 2^d rows.

Reply:
perhaps i'm reading it wrong, that happens quite often :-) - but isn't that what `explode` _would_ do, if you could pass it >1 array at once ? (perhaps minus the permutations with `null` for one or more inputs)

Reply:
happy to be wrong, perhaps a richer example would help?

Reply:
Hey @blueedgenick! These null permutations is exactly what we want. I added an example and some links. Hope this makes it more clear.